compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64MNegTest.java

```diff
@@ -0,0 +1,143 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package org.graalvm.compiler.core.aarch64.test;
+
+import java.util.function.Predicate;
+import org.graalvm.compiler.lir.LIRInstruction;
+import org.junit.Test;
+
+public class AArch64MNegTest extends AArch64MatchRuleTest {
+    private static final Predicate<LIRInstruction> predicate = op -> (op.name().toUpperCase().equals("MNEG"));
+
+    private void testAndCheckLIR(String methodName, int insCount, Object... args) {
+        assert args.length >= 2;
+        test(methodName, args);
+        checkLIR(methodName, predicate, insCount);
+    }
+
+    /**
+     * mneg rule test with int type.
+     */
+    private static int mulNegInt(int x, int y) {
+        return ((x + 1) * 2 + 3) * (-y);
+    }
+
+    private static int mulZeroSubInt(int x, int y) {
+        return x * (0 - y);
+    }
+
+    private static int mulSubZeroInt(int x, int y) {
+        return x * (y - 0);
+    }
+
+    public static int multiplyNegateInt(int x, int y) {
+        return mulNegInt(x, y) + mulZeroSubInt(y, x) + mulSubZeroInt(x, y);
+    }
+
+    @Test
+    public void testMultiplyNegateInt() {
+        testAndCheckLIR("multiplyNegateInt", 2, 7, -1);
+    }
+
+    public static int negMulInt(int x, int y) {
+        return (-((x + 1) * y));
+    }
+
+    @Test
+    public void testNegateMultiplyInt() {
+        testAndCheckLIR("negMulInt", 1, 7, -1);
+    }
+
+    /**
+     * mneg rule test with long type.
+     */
+    private static long mulNegLong(long x, long y) {
+        return ((x + 1) * 2 + 3) * (-y);
+    }
+
+    private static long mulZeroSubLong(long x, long y) {
+        return x * (0 - y);
+    }
+
+    private static long mulSubZeroLong(long x, long y) {
+        return x * (y - 0);
+    }
+
+    public static long multiplyNegateLong(long x, long y) {
+        return mulNegLong(x, y) + mulZeroSubLong(y, x) + mulSubZeroLong(x, y);
+    }
+
+    @Test
+    public void testMultiplyNegateLong() {
```

Comment:
After merging 77d0221b486eac09756a26144632ee04fc3afc67, this test fails because the second MNEG is concurring with the MADD rule. I am updating the test accordingly.

Reply:
I would even say, just to test the single case with `mulNegLong`, since we assume that the compiler can optimize `(0-y)` to `-y`.